### ISSUE TITLE
converted exported material no edge, no shadow, no editable

### ISSUE THREAD

**Issue by Hogarth-MMD** [powroupi/blender_mmd_tools#109]
https://sta.sh/01zyyhmhpqw

Hello @powroupi . This is again a link to Sintel butterflies. These models have an issue. One of my earlier Sintel exports has this issue also. One or more materials are unable to cast a shadow in MMD. They are also not able to have edge lines around the material in MMD. In MikuMikuMoving, however, there is no problem with casting shadows. Even if I correct the edge and shadow settings for these materials in PMD editor and PMX editor, the issue is not corrected. 
These models are so damaged by some strange, defective export by mmd_tools that they cannot even be corrected and edited in PMD editor or PMX editor. I have wasted many hours trying to troubleshoot this issue, and I have decided that there is a  95% probability that this is being caused by mmd_tools. This is a very strange, seemingly impossible issue. How can mmd_tools damage a pmx model in a way that it cannot be edited and corrected in PMD editor or PMX editor.? This is a mysterious anomaly. Will you please explain to me how this happens and will you please fix this issue.

**Comment by nathanvasil** [comment]
@Hogarth-MMD I looked at butterfly orange.pmx.  It doesn't have Shadow or Edge Line enabled for any of its materials.  After enabling edge line for all materials, most of the model has an edge line.  The wings don't have an edge line because these are flat, planar meshes-- it's drawing an edge line, but because the edges of the mesh are discontinuous, when it uses those normals to create an edge, the edge always lies behind the mesh, so you never can see the edge.  (This makes more sense if you understand how MMD's edge line is created.  To make an edge, it draws the mesh twice.  Once to draw it normally, and once for the edge line.  For the edge line, it reverses the directions of all of the faces and translates all the vertices along their original normals in an amount that depends on edge line thickness, distance, and perhaps FOV.)

Not sure what's happening with shadow yet.  I'll let you know when I figure it out.


Edit: NM, it will still draw an edge here, although you'll see it's kind of crappy.  It's just that in addition to edge line disabled, these models have an edge line Alpha of 0 set for the wings.

**Comment by nathanvasil** [comment]
Okay, I can confirm that shadow does work once you enable it on the models.

![bugshadow](https://user-images.githubusercontent.com/19930899/35763823-982b793a-0869-11e8-9930-c94f4fbfd6bd.jpg)

To make sure you're exporting models with edge line and shadow casting enabled, select a material and look at the MMD Material subsection of the materials tab of the properties view.  The relevant properties here are "Toon Edge" (MMDMaterial.enabled_toon_edge) and "Self Shadow"  (MMDMaterial.enabled_self_shadow).

Edit: Oh, no wing shadows.  I believe this is because you're using a texture with alpha for the wings.  When you enable edge line, it does draw shadows.  Never knew edge lines cast shadows or that alpha textures wouldn't, learn something new all the time.

Edit2: Oh, NM.  Edge line color determines shadow color.  Since your wings have an alpha of 0, it's not casting any shadow.  If you change this to 1, even without enabling edge line, it will cast shadows appropriately.  There are no problems with it being an alpha-mapped texture (although the shadows won't respect the alpha).

**Comment by Hogarth-MMD** [comment]
Thank you for your thoughts, @nathanvasil , but much of what you are saying is simply incorrect. On an exported model of Sintel herself, every material is not casting a shadow except her hair. The other materials are not flat and not transparent. Even after the texture is removed from a material and the alpha is set to 1 in PMD editor or PMX editor, the problem remains. This issue seems to happen with every material which has a texture image. And even after every conceivable fix in PMD editor, or PMX editor the issue is not fixed. Obviously when converting a model, someone should not need to manually enable edge line on every material.

**Comment by Hogarth-MMD** [comment]
"Edge line color determines shadow color."   
I am observing this also. I think that this is an astute and correct observation which you have made. But this leaves unexplained why this cannot be corrected after export in PMD editor or PMX editor.

**Comment by nathanvasil** [comment]
If you want me to figure out a different model, I'd take a look.  To fix the butterfly model you've linked here, enable self shadow, edge line, and change the edge color alpha to 1.  My testing shows that working in MMD.

Shadow values are probably not populated on export from MMD tools, and maybe ought to be.  Edge line, it wouldn't make any sense-- how can Blender know what you want to add an edge to?

"cannot be corrected after export in PMD editor or PMX editor."

It can be corrected.  I used PMX editor to correct it, not Blender.

**Comment by Hogarth-MMD** [comment]
In PMX editor and PMD editor you can create a plane primitive object. I just did several experiments with a plane primitive object. It has no problem casting a shadow in MMD. No adjustment of edge settings in PMD editor has any effect on its casting of a shadow or its color. There does seem to be some kind of cause and effect relationship with edge line and shadow, but I am not sure what exactly that consists of .

**Comment by Hogarth-MMD** [comment]
Finally I have success displaying shadow of the butterfly wings in MMD.  Changing the alpha value of the edge line from 0 to 1 fixes this issue. This is the crucial step which I omitted. There is no dependence of the shadow color on the edge line color, but there is a dependence of the shadow on the edge line alpha (in MMD but not in MMM). This is excellent information @nathanvasil and I thank you very much for this excellent information.

**Comment by powroupi** [comment]
Currently, the conversion of MMD material is very simple, see [operators/model.py](https://github.com/powroupi/blender_mmd_tools/blob/5aeb4c62062ffa6ce568e76f3dfb4b06b6bec535/mmd_tools/operators/model.py#L212-L221). So, don't expect the materials will be converted perfectly. But I think nothing is read only, you should be able to adjust MMD materials in Blender and PMDE/PMXE. :smile: 

And thanks for your discussion, I didn't know the issue of edge line and shadow. :blush:

**Comment by Hogarth-MMD** [comment]
\operators\model.py
ConvertToMMDModel(Operator)

Hello @powroupi! This bug should be fixed. This is an extremely simple bug fix. Do you want me to send this bug fix to you as a pull request?
We have 2 choices in fixing this bug. We can use m.line_color or not use m.line_color . Which of these 2 choices do you prefer? I looked through the Blender interface and in the Blender interface, I cannot find any line_color attribute of a material, but this line_color attribute of a material does exist in Python.

mmd_material.edge_color[3] = 1.0

OR

m.line_color[3] = 1.0
mmd_material.edge_color = m.line_color

**Comment by powroupi** [comment]
Maybe 
```python
if mmd_material.edge_color[3] < 0.1:
    mmd_material.edge_color[3] = 1.0
```
is better I think. I can fix it in next update. :smile:

**Comment by nathanvasil** [comment]
Well, you do have the potential to screw up MMD models that are using this intentionally to control shadows, since you can create material morphs that affect edge line color (incl. alpha), but you can't create morphs that alter material flags.

If I understand correctly, this is only an issue for Hogarth due to exporting models not intended for MMD.  What happens if a user doesn't specify an edge line color on exporting a model?  Does it export as 0,0,0,0?  That seems like a bad idea for anything set to cast shadows.  But on the other hand, if this is only an issue when a user has specifically entered a 0 alpha edge line, then shouldn't MMD Tools behave as if the user knows what he or she is doing, rather than overriding what they tell it to do?

**Comment by Hogarth-MMD** [comment]
This is just a bug fix for the Convert Model operator. Nothing else would be affected.

**Comment by powroupi** [comment]
Default value of MMD material `edge_color` is [0, 0, 0, 1], and toon edge is disabled, so there won't be any issue if a user doesn't specify an edge line color on exporting a model.

Default value of Blender Freestyle `line_color` is [0, 0, 0, 0], so it may cause issues when converting a non-MMD model via Convert Model operator. We are going to fix this issue. :smile:

**Comment by nathanvasil** [comment]
Okay, sorry that I misunderstood what was going on.

**Comment by powroupi** [comment]
I think the issue is solved in commit 89b850e. :yum:
